### DISCUSSION TITLE
Best way to add an event listener to save resources

### DISCUSSION THREAD

**Discussion by shunz19** [socketio/socket.io#3771] [Q&A] [answered]
So i'm trying to know how javascript handles function.

if i have a code like this:
```javascript
io.on("connection", function(socket) {
  socket.on("hi", function(data) {
    socket.emit("emit", "hey")
  })
})
```
What's in my mind is that each new connection, javascript will create another function for the "hi" event. So what I'm currently doing in my app is like this:
```javascript
function hi(data) {
    this.emit("emit", "hey")
}

io.on("connection", function(socket) {
  socket.on("hi", hi)
})
```
This way javascript will just reuse the function hi instead of instancing a new one?? I'm not sure if this is necessary but I want to consume less resources as much as possible because I'm thinking about what will happen when the server have thousand of connections.

If what i'm currently doing is correct, how do I implement this style without my code turning into spaghetti? Or is it really worth it to use this kind of coding to save memory resources?

**Comment by darrachequesne** [comment, marked as answer]
Here's a common application structure:

`index.js`

```js
const httpServer = require("http").createServer();
const io = require("socket.io")(httpServer);

const { createOrder, readOrder } = require("./orderHandler")(io);
const { updatePassword } = require("./userHandler")(io);

const onConnection = (socket) => {
  socket.on("order:create", createOrder);
  socket.on("order:read", readOrder);

  socket.on("user:update-password", updatePassword);
}

io.on("connection", onConnection);
```

`orderHandler.js`

```js
module.exports = (io) => {
  const createOrder = function (payload) {
    const socket = this; // hence the 'function' above, as an arrow function will not work
    // ...
  };

  const readOrder = function (orderId, callback) {
    // ...
  };

  return {
    createOrder,
    readOrder
  }
}
```

I've added it here: https://socket.io/docs/v3/server-application-structure/

**Reply by shunz19** [reply]
Hi, check out this code:
```javascript
var a = function() {
    return 1;
}
var b = a;
// b === a is true

var socket = {}
var c = a.bind(socket);
// b === c is false
```
As far as I'm aware, added listener functions are binded with the socket.
Will the javascript create a new function when using .bind??

**Reply by darrachequesne** [reply]
According to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind

> The `bind()` method creates a new function that, when called, has its `this` keyword set to the provided value, with a given sequence of arguments preceding any provided when the new function is called.

That being said, if your use case is that memory sensitive, maybe Javascript isn't the best option :smiley_cat:
